### ISSUE TITLE
Flux-forced simulations - varying heat flux by a constant

### ISSUE THREAD

**Issue by dhruvbhagtani** [dhruvbhagtani/sfc-perturbation-expts#40]
Unlike #39, here, we change the surface heat fluxes on the globe by a certain factor. The idea is to add heat to subtropics and lose out heat from the subpolar regions. So, in a way, not only are we changing the magnitude of heat flux, we are also changing the spatial structure. As a starting point, we aim to change the heat fluxes by 20 W/m2. Some potential issues:

1. How do we identify the two regions of heat gain and heat loss? Some possible suggestions are: (i) Using barotropic stream function, and (ii) Using SSH contours.
2. Once the two regions of heat gain and heat loss are identified, how do we apply the smoothing function to ensure that we don't have step functions for surface heat fluxes anywhere?
3. Finally, we need to make sure that the heat gain and heat loss are balanced, so that the ocean doesn't have a net non-zero heat flux.

**Comment by dhruvbhagtani** [comment]
This comment is for deciding how to identify the two regions of heat gain and heat loss.

Here is a preview of how the barotropic streamfunction and SSH contours across the globe look like. Note that in the stream function plot, the red contours represent the subtropical gyres in both hemispheres, and vice versa. The area enclosed by dotted region is where heat loss will occur, and the area enclosed by solid line is where heat gain will occur.

We exclude the tropical region from this analysis. Then, except the eastern Atlantic (both subtropical and subpolar regions), the two plots will give us roughly similar regions of heat gain and heat loss. So, let's focus on the eastern Atlantic. I feel that the barotropic streamfunction is a more accurate representation of where gyres are supposed to be, and so choosing it makes more sense. What do you all think?

![psi](https://user-images.githubusercontent.com/29700296/153996827-86d6f31a-4b30-4aaf-8b97-9b4f469f951e.jpeg)

![ssh](https://user-images.githubusercontent.com/29700296/153996841-f294b248-7b90-48b9-ab83-9e76cf097651.jpeg)

**Comment by navidcy** [comment]
I vote for stream function. 

How would you make the forcing function so that it saturates to +/-1 near the ψ=0?

**Comment by navidcy** [comment]
What do you mean "exclude tropics"? From the plots pr from the perturbation exps?

**Comment by rmholmes** [comment]
I think you need to try a few functions (of SSH and BT stream-function) and see what it looks like. It looks to me like you're going to run into problems in the Atlantic (particularly at low latitudes) with both - but you won't know until you try it.

**Comment by dhruvbhagtani** [comment]
@navidcy I meant we should exclude tropics rom the plots. But now that I think of it, we do need to discuss what we should do with the tropics. I am inclined towards keeping the fluxes constant. Certainly not in favour of reducing the heat fluxes, as the streamfunction plot suggests.

Also, to answer the next question about how we'll make the forcing, I am thinking of some sort of smoothing function. An example is this: https://math.stackexchange.com/questions/417717/looking-for-a-3d-smooth-step-function. But I am still working on it.,

**Comment by dhruvbhagtani** [comment]
I suggest an approach to create our wind mask, which is as follows (I refer to locations of net heat input as HI and locations of net heat output as HO):
1. 1st global conditions for HI: abs(latitude) should at least be 40 degrees N/S.
2. 2nd global condition for HI: abs(latitude) cannot be greater than 45 degrees N/S.
3. 1st local condition: For Indian and Pacific oceans, SSH > 0.1 m for HI.
4. 2nd local condition: For Atlantic oceans, SSH > -0.3 m for HI.

![heat_map](https://user-images.githubusercontent.com/29700296/154413705-19ab0ea0-b046-4be3-9ddf-5dfabb54d896.jpeg)


I think tan hyperbolic function would be better for y-axis interpolation at the junction between heat gain and heat loss regions. Since it is a bit more complicated, for now, I'm just showing linear interpolation.

![interpolation](https://user-images.githubusercontent.com/29700296/154413736-f3b0187c-341b-4faf-903c-2222f4a28611.jpeg)

Finally, I plot the junction (where mask = 0) as a contour on the streamfunction plot to check its sanity. I feel that the mask does a good job to capture the subtropical and subpolar gyres clearly in the Northern Hemisphere, but no so much in the Southern Hemisphere. For that, I guess I'll have to increase the latitude in condition 2. 

![heat_map_over_psi](https://user-images.githubusercontent.com/29700296/154413856-2becac51-468f-41cd-9f04-49c17889d43a.jpeg)

**Comment by navidcy** [comment]
- why the mask's contour at the last plot does not coincide with the zero-contour of the Barotropic stream function in the Southern Ocean?
- perhaps not important but a your masks introduces sharp gradients... I mean, the derivative of your mask has discontinuities! I'd always be in favor of smoothing things out with `tanh` functions...

**Comment by rmholmes** [comment]
Did you decide that a tanh function of SSH wasn't going to work?

**Comment by navidcy** [comment]
Also it seems that you mask is only applied in lat... So there is no smoothing in longitude?
(I'm just looking near South Africa for example)

**Comment by navidcy** [comment]
Why did you need to have those four conditions with ad-hoc latitudes? What happens if you drop them?

(Sorry if I'm asking things that you somehow asked yourself and came to a conclusion somehow.)

**Comment by AndyHoggANU** [comment]
Yeah, so, if you can make the zero-line smooth, then I don’t think you need smoothing in the zonal direction. If not, then you may need a little zonal smoothing.


I am also slightly concerned about the latitude in the Southern Hemisphere. I feel like it should be slightly further south, even if that means choosing different latitude ranges for the southern and northern cutoff values.

@rmholmes - the reason for avoiding tanh of SSH is that SSH anomalies are very small near the Easter boundaries, yet I was arguing this is still part of the gyre (i.e. anomalies there will propagate westwards into the gyre). The more zonal we can make this forcing, the better, I think.

Lastly, do we want to ensure zero net heat input?

**Comment by rmholmes** [comment]
Sounds good to me. I agree that the SH transition is a bit funny - you want the subtropical gyres to experience anomalous heating, not a transition between the two. The 0 line of the BT streamfunction seems like it would work reasonably well for this?

I think ensuring zero net heat input is important.

**Comment by dhruvbhagtani** [comment]
Yes, I think the mask needs some more work. Answering some questions:
1. The mask's contour doesn't coincide with the zero-contour of the Barotropic stream function in the Southern Ocean because we are not taking 0 Sv as the contour, instead we are taking contours using SSH. But as Andy pointed out, instead of worrying about symmetry, I should make the mask in a way which captures the southern ocean gyres properly.
2. Yep, the derivatives will have discontinuities, so this is something which can be solved using a smoothing function, which I am working on.
3. No, I haven't ruled out the tanh smoothing of SSH. However, even if we do that, it will only be along y-axis, not along x-axis. 
4. Since the anomalies will propagate westward, it is crucial that we apply anomalies over the entire longitude range of the basin, not only where the gyre is present.

**Comment by dhruvbhagtani** [comment]
You're right @rmholmes, the next task will be to ensure net zero heat.

**Comment by dhruvbhagtani** [comment]
Here is the updated map with tanh smoothing. I have smoothed out the contour in the Southern Ocean a bit too. 

![heat_map_tanh](https://user-images.githubusercontent.com/29700296/155065080-64b83b36-2b12-45e3-9a48-a15b52b71084.jpeg)
![tanh_one_lon-2](https://user-images.githubusercontent.com/29700296/155065102-02cdd26a-d913-43b4-9576-d4df9147e056.jpeg)


The smoothing is done over about 12 degrees latitude (i.e., transition from +1 to -1 occurs over 12 degrees). We can increase it even further (streamfunction map is present just below for reference).

![heat_map_tanh_over_psi](https://user-images.githubusercontent.com/29700296/155065090-c5fa10d3-5f44-4681-b989-cbfda37dc9c3.jpeg)


I am now starting to move on to how to keep net zero heat flux, but if you have any suggestions from this plot, I'm happy to change it.

**Comment by navidcy** [comment]
I like that figure.... there are still some discontinuities south of Cape of Good Hope but that's OK I think.

If you compute the area integral of the positives and the negatives how much is the imbalance?

Like what's the area-integral of the positives heatflux perturbation ∫`Θ(mask)` dx dy 
compared to the the sum of the positives and negatives

( ∫`Θ(mask)` dx dy +  ∫`Θ(-mask)` dx dy) ,

where `Θ(x)` is

```
def Θ(x):
    return 1/2 * np.sign(x) + 1/2
```

**Comment by navidcy** [comment]
```python
import numpy as np
import matplotlib.pyplot as plt

def Θ(x):
    return 1/2*np.sign(x)+1/2

x = np.linspace(-10, 10, 100)

plt.plot(x, Θ(x));
```
![Unknown](https://user-images.githubusercontent.com/7112768/155065948-df4864af-f733-495f-8031-6c322d88af59.png)

**Comment by navidcy** [comment]
(probably there is an easier way... I just wanted to mask everything that is positive or negative...)

**Comment by navidcy** [comment]
> Here is the updated map with tanh smoothing. I have smoothed out the contour in the Southern Ocean a bit too.
> 
> ![heat_map_tanh](https://user-images.githubusercontent.com/29700296/155065080-64b83b36-2b12-45e3-9a48-a15b52b71084.jpeg) ![tanh_one_lon-2](https://user-images.githubusercontent.com/29700296/155065102-02cdd26a-d913-43b4-9576-d4df9147e056.jpeg)
> 
> The smoothing is done over about 12 degrees latitude (i.e., transition from +1 to -1 occurs over 12 degrees). We can increase it even further (streamfunction map is present just below for reference).
> 
> ![heat_map_tanh_over_psi](https://user-images.githubusercontent.com/29700296/155065090-c5fa10d3-5f44-4681-b989-cbfda37dc9c3.jpeg)
> 
> I am now starting to move on to how to keep net zero heat flux, but if you have any suggestions from this plot, I'm happy to change it.

I'd try smoothing over double the range..., e.g., from -1 to 1 over 24 degrees? Just to see how it looks at least.

**Comment by dhruvbhagtani** [comment]
I'll try out the wider smoothing.

Regarding your other question: ratio of +ve to -ve area is 3:1. Didn't expect this. So, if we were to heat the red area uniformly by 20 W/m2, we'll have to cool the blue area by 60 W/m2!

**Comment by navidcy** [comment]
Why you didn't expect this? This is a projection. Area at the poles seems bigger than in reality... :)

**Comment by dhruvbhagtani** [comment]
Maybe I was looking at the map without any continents and expected a 1.5:1 ratio at max. But anyways, we'll have to deal with it. I will think about it. Certainly not in favour of heating by 20 and cooling by 60 W/m2.

**Comment by navidcy** [comment]
>  Certainly not in favour of heating by 20 and cooling by 60 W/m2.

Why not?

**Comment by navidcy** [comment]
perhaps if we exclude heading in the tropics then the heating-cooling could be equal?

**Comment by dhruvbhagtani** [comment]
> > Certainly not in favour of heating by 20 and cooling by 60 W/m2.
> 
> Why not?

Two reasons:
1. Cooling the subpolar and polar gyres by such a huge factor could lead to a very weird overturning response.
2. We already have a sort of temperature restoring in the Arctic and Antarctic through frazil formation. By increasing the heat loss, we can expect a very strong and unrealistic response of frazil in these polar regions.

Yep, I think not heating (or maybe cooling to some extent) the tropics would help.

**Comment by AndyHoggANU** [comment]
OK, so another possibility is to taper your heating function off as you approach the Equator. This will change the 3:1 ratio quite a bit and might make +/-20 doable?

Otherwise, map looks good to me.

**Comment by dhruvbhagtani** [comment]
Yes, I have done something on similar lines - did you mean something like this?

![one_latitude_corrected](https://user-images.githubusercontent.com/29700296/155272140-568ece7c-39f3-46c3-8aca-f7e56f1fe4d9.jpeg)

If that's correct, here is how the contour (of 0 W/m2 heating) looks like on a streamfunction plot:

![heat_map_over_psi_corrected](https://user-images.githubusercontent.com/29700296/155272319-c8d82470-2911-4132-9b3b-6e81ec960ab2.jpeg)

Another good thing is that the average value of our heat flux map value is O(10^(-9)), so more or less, the heat flux imbalance is also sorted.

**Comment by dhruvbhagtani** [comment]
I think you 

> OK, so another possibility is to taper your heating function off as you approach the Equator. This will change the 3:1 ratio quite a bit and might make +/-20 doable?
> 
> Otherwise, map looks good to me.

If you meant that we taper if off continuously over the subtropical gyre, I would rather prefer tapering it off sharply near the equator (like I have done above), reason being that we want our gyres to experience the same heating throughout their spatial structure. What do you think?

**Comment by dhruvbhagtani** [comment]
> OK, so another possibility is to taper your heating function off as you approach the Equator. This will change the 3:1 ratio quite a bit and might make +/-20 doable?
> 
> Otherwise, map looks good to me.

If you meant that we taper if off continuously over the subtropical gyre, I would rather prefer tapering it off sharply near the equator (like I have done above), reason being that we want our gyres to experience the same heating throughout their spatial structure. What do you think?

**Comment by dhruvbhagtani** [comment]
> OK, so another possibility is to taper your heating function off as you approach the Equator. This will change the 3:1 ratio quite a bit and might make +/-20 doable?
> 
> Otherwise, map looks good to me.

If you meant that we taper if off continuously over the subtropical gyre, I would rather prefer tapering it off sharply near the equator (like I have done above), reason being that we want our gyres to experience the same heating throughout their spatial structure. What do you think?

**Comment by dhruvbhagtani** [comment]
Yep, I tapered off the heating function as we approach the equator, again using tanh function. It now looks like:

![one_latitude_corrected](https://user-images.githubusercontent.com/29700296/155653245-1bd0849c-e9fc-48d5-9331-e2ea7911ee57.jpeg)

The overall map looks like (where the black contour represents 0 heat gain/loss):
![heat_map_corrected](https://user-images.githubusercontent.com/29700296/155653302-cced8e05-0065-4ea9-ae22-39dea7c07642.jpeg)

This same black contour over the streamfunction plot is:
![heat_map_over_psi_corrected](https://user-images.githubusercontent.com/29700296/155653390-00ed3ee1-b739-4e3d-a0ad-34b9fc564429.jpeg)

A_pos -> Total area where mask = +1
A_neg -> Total area where mask = -1
By not having anything in the equator, we have managed to reduce A_pos/A_neg ratio from 3:1 to 1.5:1. Now, we have two options:
1. Have a small negative (~-0.25) heat flux at the equator where currently we have 0.
2. Continue with this map and change A_neg region from -1 to -1.5. 

I am more inclined towards doing the first one because we have a lot of heat gain near the equator, and by subtracting a small amount (5-6 W/m2) should not make much difference. On the other hand, losing more heat at the poles could lead to more frazil formation in our ocean-only model, which I am not a big fan of.

**Comment by navidcy** [comment]
Go with your intuition @dhruvbhagtani.

**Comment by dhruvbhagtani** [comment]
The new mask looks like this:

![heat_map_over_psi_corrected](https://user-images.githubusercontent.com/29700296/156511830-ed54420b-d387-4875-a52b-d5f0c52677f8.jpeg)

![heat_map_corrected](https://user-images.githubusercontent.com/29700296/156511821-f678a21e-4de6-42c1-bc8a-bee0024dc406.jpeg)

![one_latitude_corrected](https://user-images.githubusercontent.com/29700296/156511788-42cde9e6-14b0-4bb0-bc1d-f107d765058b.jpeg)

Negative mask: -1
Positive mask: 0.56.

With this, the net heat input in the ocean is zero.

**Comment by navidcy** [comment]
@dhruvbhagtani did you post something and unpost it? I got an email but I can't see it anymore here.

**Comment by AndyHoggANU** [comment]
Yeah, me too. What came through in the email looked good, but I couldn't find it on github.

**Comment by navidcy** [comment]
I had questions on that -- but perhaps also @dhruvbhagtani had second thoughts when he saw it and decided to retract :)

**Comment by dhruvbhagtani** [comment]
Oh, sorry. There is an issue with the mask; it starts to taper off in the subtropical gyre region only. And I haven't been able to figure it out because of OSM. Will get onto it after OSM.

**Comment by dhruvbhagtani** [comment]
Okay, so the issue with mask has been resolved, and we are continuing with no heat gain in the equatorial band, 0.5x heat gain in the tropical-subtropical band, and 1x heat loss in the subpolar polar band. This also ensures that we don't have any additional heat input in the ocean.

**Comment by dhruvbhagtani** [comment]
Here is the GitHub link for the control MOM5 experiment (https://github.com/dhruvbhagtani/varying-surface-forcing/blob/main/025deg_flux_forced_uniform/Control.ipynb), which is spun off from the ACCESS-OM2 fluxes (created from the last 20 years of the ACCES-OM2 run). Some comments/questions:
1. I'm a bit concerned about the surface temperature increasing which can induce unnecessary density perturbations which aren't linked to heat fluxes, and so it might become a problem. What do you all think?
2. When we picked the last 20 years from ACCESS-OM2 control, I was quite satisfied with the way AMOC and AABW metrics were looking. But AABW has reduced drastically in the 100 years of the MOM5 control. I hope it's not a big issue, since 0.25 degree models can't model AABW accurately anyways.
3. I am continuing the MOM5 control for another 20 years, since I don't think we have reached a state of equilibrium. But again, running it more means that we are continuously adding heat into the ocean, which will increase our SSTs even more. So, will we ever be able to reach equilibrium with these set of parameters?
4. I think the gyres have reached a state of equilibrium, and since we are concerned only with them, we can instead choose not to run this control experiment any longer and apply the perturbations, but this time I would like to ensure that we don't have any net heat gain.

**Comment by dhruvbhagtani** [comment]
Finally we got the results we wanted!
Comparing two experiments here, (i) Control, and (ii) Heat perturbation with 20 W/m2 heat loss at higher latitudes and 10 W/m2 heat gain at lower latitudes. Longitudinally averaged heat flux maps show an increased heat contrast, which we ideally wanted:
![Screen Shot 2022-04-13 at 4 04 23 pm](https://user-images.githubusercontent.com/29700296/163110802-053cebd0-ffb7-4915-aad0-c087261848b4.png)

The North Atlantic gyres (integrated to different depths) shows an increased circulation in (ii). 
1. Control
<img width="1229" alt="Screen Shot 2022-04-13 at 3 57 45 pm copy" src="https://user-images.githubusercontent.com/29700296/163110962-2c95749a-c442-4afc-9b63-e377a89d8b2e.png">
2. Heat perturbation experiment
<img width="1229" alt="Screen Shot 2022-04-13 at 3 57 45 pm" src="https://user-images.githubusercontent.com/29700296/163110989-67a8a099-60af-4b4b-87f3-a4c0105d7e10.png">
Other experiments are still ongoing.

**Comment by AndyHoggANU** [comment]
Looks pretty clear!!

**Comment by dhruvbhagtani** [comment]
These are the percentile-time series area-weighted plots for different heat perturbation experiments.

1. Green line is $\psi$ = 30 Sv.
<img width="987" alt="Screen Shot 2022-05-31 at 3 15 31 pm" src="https://user-images.githubusercontent.com/29700296/171098019-ce2aa907-0d30-47ad-9243-0fc01f41be56.png">
2. Green line is $\psi$ = 25 Sv.
<img width="1005" alt="Screen Shot 2022-05-31 at 3 15 14 pm" src="https://user-images.githubusercontent.com/29700296/171098285-1ab2ece7-8053-4664-81d4-88d29fe047c0.png">

Very intrigued by the $\psi$ = 25 Sv line, as it shows a significant increase in heat addition experiments in the beginning.

**Comment by dhruvbhagtani** [comment]
By intrigued, I mean "happy intrigued", as this variability is precisely what I observed in animations as well.

**Comment by dhruvbhagtani** [comment]
The figures show the first vs last 18 years of our perturbation experiments.

1. Wind stress perturbations
![Last_18yrs_wind](https://user-images.githubusercontent.com/29700296/172099227-badc1e68-19bc-4d68-ac15-d1c7a240e58b.png)

2. Buoyancy flux perturbations
![Last_18yrs_heat](https://user-images.githubusercontent.com/29700296/172099238-e7b33921-4c05-4eab-b814-39e04bb3f83b.png)

This is in line with the above figures too, the non-linear effects start to dominate after some time.
